include/morum/tree_node.hpp

```diff
@@ -0,0 +1,353 @@
+/**
+ * Copyright Quadrivium LLC
+ * All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <utility>
+
+#include <qtils/bytes.hpp>
+#include <qtils/cxx23/forward_like.hpp>
+#include <qtils/optional_ref.hpp>
+
+#include <morum/common.hpp>
+
+namespace morum {
+
+  struct TreeNode;
+
+  // an identifier for a tree node in memory
+  using NodeId = uint64_t;
+
+  /**
+   * An extended tree branch, where a child may be stored as either a subtree
+   * hashe or a NodeId. Children stored by NodeId are supposedly loaded to RAM
+   * because they are located on modified paths and need to be re-hashed.
+   */
+  struct Branch {
+    friend class MerkleTree;
+    friend TreeNode deserialize_node(qtils::ByteView bytes);
+
+    static constexpr NodeId NoId = static_cast<NodeId>(-1);
+    static constexpr Hash32 NoHash = []() {
+      auto h = ~ZeroHash32;
+      h[0] &= 0xFE;  // first bit is always zero for branch nodes
+      return h;
+    }();
+
+    explicit Branch(const Hash32 &left, std::nullopt_t)
+        : left_hash{left}, right_hash{NoHash}, left_idx{NoId}, right_idx{NoId} {
+      left_hash[0] &= 0xFE;  // the first bit in a branch must be zero
```

Comment:
```suggestion
      left_hash[0] &= 0b1111110;  // the least bit in a branch must be zero
```
and the same for such cases

Reply:
I agree with the first -> least change, but the I'd rather leave the literal in hex. It's going to be replaced with unsetLeastBit in most places anyway, as per your suggestion.